File: packages/openclaw/src/monitor/history.ts
import type { ClientPostBlobData } from '@tloncorp/api';
import type { RuntimeEnv } from 'openclaw/plugin-sdk/runtime';

import { formatBlobForHistory, parseBlobData } from './media.js';
import { extractMessageText } from './utils.js';

/**
 * Format a number as @ud (with dots every 3 digits from the right)
 * e.g., 170141184507799509469114119040828178432 -> 170.141.184.507.799.509.469.114.119.040.828.178.432
 */
function formatUd(id: string | number): string {
  const str = String(id).replace(/\./g, ''); // Remove any existing dots
  const reversed = str.split('').toReversed();
  const chunks: string[] = [];
  for (let i = 0; i < reversed.length; i += 3) {
    chunks.push(
      reversed
        .slice(i, i + 3)
        .toReversed()
        .join('')
    );
  }
  return chunks.toReversed().join('.');
}

export type TlonHistoryEntry = {
  author: string;
  content: string;
  timestamp: number;
  id?: string;
  blob?: string | null;
  parsedBlobData?: ClientPostBlobData | null;
};

export const MAX_THREAD_CONTEXT_MESSAGES = 20;

type ParentPostEssay = {
  author?: string | { ship?: string };
  content?: unknown;
  sent?: number;
};

type ParentPostSeal = {
  id?: string;
};

type ReplyMemo = ParentPostEssay & {
  blob?: string | null;
};

type ExactReplyScryResponse = {
  reply?: {
    memo?: ReplyMemo;
    'reply-essay'?: ReplyMemo;
    seal?: ParentPostSeal;
  };
  memo?: ReplyMemo;
  'reply-essay'?: ReplyMemo;
  seal?: ParentPostSeal;
};

function normalizeMessageId(id: string | number | undefined | null): string {
  const rawId = String(id ?? '');
  const separatorIndex = rawId.indexOf('/');
  const author = rawId.slice(0, separatorIndex);
  const bareId =
    separatorIndex > 0 && /^~?[a-z0-9-]+$/i.test(author)
      ? rawId.slice(separatorIndex + 1)
      : rawId;
  return bareId.replace(/\./g, '');
}

function getParsedBlobData(entry: TlonHistoryEntry): ClientPostBlobData | null {
  if (entry.parsedBlobData !== undefined) {
    return entry.parsedBlobData;
  }

  entry.parsedBlobData = parseBlobData(entry.blob);
  return entry.parsedBlobData;
}

/**
 * Render a history entry's full content for context display.
 * Combines text content with compact blob annotations (e.g. voice memo transcripts).
 */
export function renderHistoryContent(entry: TlonHistoryEntry): string {
  const parts: string[] = [];
  const blobData = getParsedBlobData(entry);
  if (blobData) {
    const blobText = formatBlobForHistory(blobData);
    if (blobText) parts.push(blobText);
  }
  if (entry.content) parts.push(entry.content);
  return parts.join('\n');
}

const messageCache = new Map<string, TlonHistoryEntry[]>();
const reactionTargetCache = new Map<string, TlonHistoryEntry[]>();
const MAX_CACHED_MESSAGES = 100;
const MAX_CACHED_REACTION_TARGETS = 20;
const REACTION_TARGET_FETCH_ATTEMPTS = 3;
const REACTION_TARGET_FETCH_RETRY_DELAY_MS = 50;

function findCachedMessage(
  cache: TlonHistoryEntry[] | undefined,
  messageId: string
): TlonHistoryEntry | undefined {
  if (!cache) return undefined;
  const normalizedId = normalizeMessageId(messageId);
  return cache.find(
    (entry) => entry.id && normalizeMessageId(entry.id) === normalizedId
  );
}

function removeCachedMessage(
  cache: TlonHistoryEntry[],
  messageId: string
): void {
  const normalizedId = normalizeMessageId(messageId);
  if (!normalizedId) return;

  const existingIndex = cache.findIndex(
    (entry) => normalizeMessageId(entry.id) === normalizedId
  );
  if (existingIndex >= 0) {
    cache.splice(existingIndex, 1);
  }
}

export function lookupCachedMessage(
  channelNest: string,
  messageId: string
): TlonHistoryEntry | undefined {
  return (
    findCachedMessage(messageCache.get(channelNest), messageId) ??
    findCachedMessage(reactionTargetCache.get(channelNest), messageId)
  );
}

/**
 * Resolve a channel post or reply for an event that can arrive before its echo.
 * Channel hosts assign the post id, so outbound client timestamps cannot be
 * used to pre-populate channel reaction-target lookup state.
 */
export async function lookupOrFetchCachedChannelMessage(
  api: { scry: (path: string) => Promise<unknown> },
  channelNest: string,
  messageId: string,
  rootPostId?: string,
  runtime?: RuntimeEnv
): Promise<TlonHistoryEntry | undefined> {
  const cached = lookupCachedMessage(channelNest, messageId);
  if (cached) {
    return cached;
  }

  for (let attempt = 0; attempt < REACTION_TARGET_FETCH_ATTEMPTS; attempt++) {
    const fetched = rootPostId
      ? await fetchReplyHistoryEntry(
          api,
          channelNest,
          rootPostId,
          messageId,
          runtime
        )
      : await fetchParentPostHistoryEntry(api, channelNest, messageId, runtime);
    if (fetched?.author && fetched.author !== 'unknown') {
      const echoed = findCachedMessage(
        messageCache.get(channelNest),
        messageId
      );
      if (echoed) {
        return echoed;
      }
      cacheReactionTarget(channelNest, fetched);
      return fetched;
    }
    if (attempt < REACTION_TARGET_FETCH_ATTEMPTS - 1) {
      await new Promise<void>((resolve) => {
        setTimeout(resolve, REACTION_TARGET_FETCH_RETRY_DELAY_MS);
      });
    }
  }

  return undefined;
}

export function cacheMessage(channelNest: string, message: TlonHistoryEntry) {
  if (!messageCache.has(channelNest)) {
    messageCache.set(channelNest, []);
  }
  const cache = messageCache.get(channelNest);
  if (!cache) {
    return;
  }
  removeCachedMessage(cache, message.id ?? '');
  const targets = reactionTargetCache.get(channelNest);
  if (targets) {
    removeCachedMessage(targets, message.id ?? '');
    if (targets.length === 0) {
      reactionTargetCache.delete(channelNest);
    }
  }
  cache.unshift(message);
  if (cache.length > MAX_CACHED_MESSAGES) {
    cache.pop();
  }
}

function cacheReactionTarget(
  channelNest: string,
  message: TlonHistoryEntry
): void {
  if (!reactionTargetCache.has(channelNest)) {
    reactionTargetCache.set(channelNest, []);
  }
  const cache = reactionTargetCache.get(channelNest);
  if (!cache) {
    return;
  }

  removeCachedMessage(cache, message.id ?? '');
  cache.unshift(message);
  if (cache.length > MAX_CACHED_REACTION_TARGETS) {
    cache.pop();
  }
}

export async function fetchChannelHistory(
  api: { scry: (path: string) => Promise<unknown> },
  channelNest: string,
  count = 50,
  runtime?: RuntimeEnv
): Promise<TlonHistoryEntry[]> {
  try {
    const scryPath = `/channels/v4/${channelNest}/posts/newest/${count}/outline.json`;
    runtime?.log?.(`[tlon] Fetching history: ${scryPath}`);

    const data: any = await api.scry(scryPath);
    if (!data) {
      return [];
    }

    let posts: any[] = [];
    if (Array.isArray(data)) {
      posts = data;
    } else if (data.posts && typeof data.posts === 'object') {
      posts = Object.values(data.posts);
    } else if (typeof data === 'object') {
      posts = Object.values(data);
    }

    const messages = posts
      .map((item) => {
        const essay = item.essay || item['r-post']?.set?.essay;
        const seal = item.seal || item['r-post']?.set?.seal;

        return {
          author: essay?.author || 'unknown',
          content: extractMessageText(essay?.content || []),
          timestamp: essay?.sent || Date.now(),
          id: seal?.id,
          blob: essay?.blob ?? null,
        } as TlonHistoryEntry;
      })
      .filter((msg) => msg.content || msg.blob);

    runtime?.log?.(`[tlon] Extracted ${messages.length} messages from history`);
    return messages;
  } catch (error: any) {
    runtime?.log?.(
      `[tlon] Error fetching channel history: ${error?.message ?? String(error)}`
    );
    return [];
  }
}

export async function getChannelHistory(
  api: { scry: (path: string) => Promise<unknown> },
  channelNest: string,
  count = 50,
  runtime?: RuntimeEnv
): Promise<TlonHistoryEntry[]> {
  const cache = messageCache.get(channelNest) ?? [];
  if (cache.length >= count) {
    runtime?.log?.(`[tlon] Using cached messages (${cache.length} available)`);
    return cache.slice(0, count);
  }

  runtime?.log?.(
    `[tlon] Cache has ${cache.length} messages, need ${count}, fetching from scry...`
  );
  return await fetchChannelHistory(api, channelNest, count, runtime);
}

/**
 * Fetch thread/reply history for a specific parent post.
 * Used to get context when entering a thread conversation.
 */
export async function fetchThreadHistory(
  api: { scry: (path: string) => Promise<unknown> },
  channelNest: string,
  parentId: string,
  count = 50,
  runtime?: RuntimeEnv
): Promise<TlonHistoryEntry[]> {
  try {
    // Tlon API: fetch replies to a specific post
    // Format: /channels/v4/{nest}/posts/post/{parentId}/replies/newest/{count}.json
    // parentId needs @ud formatting (dots every 3 digits)
    const formattedParentId = formatUd(parentId);
    runtime?.log?.(
      `[tlon] Thread history - parentId: ${parentId} -> formatted: ${formattedParentId}`
    );

    const scryPath = `/channels/v4/${channelNest}/posts/post/id/${formattedParentId}/replies/newest/${count}.json`;
    runtime?.log?.(`[tlon] Fetching thread history: ${scryPath}`);

    const data: any = await api.scry(scryPath);
    if (!data) {
      runtime?.log?.(`[tlon] No thread history data returned`);
      return [];
    }

    let replies: any[] = [];
    if (Array.isArray(data)) {
      replies = data;
    } else if (data.replies && Array.isArray(data.replies)) {
      replies = data.replies;
    } else if (typeof data === 'object') {
      replies = Object.values(data);
    }

    const messages = replies
      .map((item) => {
        // Thread replies use 'memo' structure
        const memo = item.memo || item['r-reply']?.set?.memo || item;
        const seal = item.seal || item['r-reply']?.set?.seal;

        return {
          author: memo?.author || 'unknown',
          content: extractMessageText(memo?.content || []),
          timestamp: memo?.sent || Date.now(),
          id: seal?.id || item.id,
          blob: memo?.blob ?? null,
        } as TlonHistoryEntry;
      })
      .filter((msg) => msg.content || msg.blob);

    runtime?.log?.(
      `[tlon] Extracted ${messages.length} thread replies from history`
    );
    return messages;
  } catch (error: any) {
    runtime?.log?.(
      `[tlon] Error fetching thread history: ${error?.message ?? String(error)}`
    );
    // Fall back to trying alternate path structure
    try {
      const altPath = `/channels/v4/${channelNest}/posts/post/id/${formatUd(parentId)}.json`;
      runtime?.log?.(`[tlon] Trying alternate path: ${altPath}`);
      const data: any = await api.scry(altPath);

      if (data?.seal?.meta?.replyCount > 0 && data?.replies) {
        const replies = Array.isArray(data.replies)
          ? data.replies
          : Object.values(data.replies);
        const messages = replies
          .map((reply: any) => ({
            author: reply.memo?.author || 'unknown',
            content: extractMessageText(reply.memo?.content || []),
            timestamp: reply.memo?.sent || Date.now(),
            id: reply.seal?.id,
            blob: reply.memo?.blob ?? null,
          }))
          .filter((msg: TlonHistoryEntry) => msg.content || msg.blob);

        runtime?.log?.(
          `[tlon] Extracted ${messages.length} replies from post data`
        );
        return messages;
      }
    } catch (altError: any) {
      runtime?.log?.(
        `[tlon] Alternate path also failed: ${altError?.message ?? String(altError)}`
      );
    }
    return [];
  }
}

async function fetchParentPost(
  api: { scry: (path: string) => Promise<unknown> },
  channelNest: string,
  parentId: string,
  runtime?: RuntimeEnv
): Promise<{ essay: ParentPostEssay; seal?: ParentPostSeal } | null> {
  try {
    // Mirrors resolveCiteContent: channels +on-peek matches `[%post time=@ ~]`, mark goes in `.json` extension.
    const scryPath = `/channels/v4/${channelNest}/posts/post/${formatUd(parentId)}.json`;
    runtime?.log?.(`[tlon] Fetching parent post: ${scryPath}`);
    const data: any = await api.scry(scryPath);

    const post = data?.post ?? data;
    const essay = post?.essay || post?.memo || post?.['r-post']?.set?.essay;
    const seal = post?.seal || post?.['r-post']?.set?.seal;
    return essay ? { essay, seal } : null;
  } catch (error: any) {
    runtime?.log?.(
      `[tlon] Error fetching parent post: ${error?.message ?? String(error)}`
    );
    return null;
  }
}

async function fetchReplyHistoryEntry(
  api: { scry: (path: string) => Promise<unknown> },
  channelNest: string,
  rootPostId: string,
  replyId: string,
  runtime?: RuntimeEnv
): Promise<TlonHistoryEntry | null> {
  try {
    const scryPath = `/channels/v4/${channelNest}/posts/post/id/${formatUd(rootPostId)}/replies/reply/id/${formatUd(replyId)}.json`;
    runtime?.log?.(`[tlon] Fetching reply: ${scryPath}`);
    const data = (await api.scry(scryPath)) as ExactReplyScryResponse | null;
    const reply = data?.reply ?? data;
    const memo = reply?.memo ?? reply?.['reply-essay'];
    if (!memo) {
      return null;
    }

    return {
      author: parentPostAuthor(memo) ?? 'unknown',
      content: extractMessageText(memo.content || []),
      timestamp: memo.sent || Date.now(),
      id: reply?.seal?.id || replyId,
      blob: memo.blob ?? null,
    };
  } catch (error: any) {
    runtime?.log?.(
      `[tlon] Error fetching reply: ${error?.message ?? String(error)}`
    );
    return null;
  }
}

function parentPostAuthor(essay?: ParentPostEssay): string | null {
  if (typeof essay?.author === 'string') {
    return essay.author;
  }
  if (typeof essay?.author?.ship === 'string') {
    return essay.author.ship;
  }
  return null;
}

/**
 * Fetch a parent post's author without requiring text content.
 */
export async function fetchParentPostAuthor(
  api: { scry: (path: string) => Promise<unknown> },
  channelNest: string,
  parentId: string,
  runtime?: RuntimeEnv
): Promise<string | null> {
  const parentPost = await fetchParentPost(api, channelNest, parentId, runtime);
  return parentPostAuthor(parentPost?.essay);
}

/**
 * Fetch the parent post body for a thread.
 */
export async function fetchParentPostHistoryEntry(
  api: { scry: (path: string) => Promise<unknown> },
  channelNest: string,
  parentId: string,
  runtime?: RuntimeEnv
): Promise<TlonHistoryEntry | null> {
  const parentPost = await fetchParentPost(api, channelNest, parentId, runtime);
  if (!parentPost) {
    return null;
  }

  const { essay, seal } = parentPost;
  const content = extractMessageText(essay.content || []);
  if (!content) {
    runtime?.log?.(`[tlon] Parent post has no text content: ${parentId}`);
  }

  const author = parentPostAuthor(essay) ?? 'unknown';

  return {
    author,
    content,
    timestamp: essay.sent || Date.now(),
    id: seal?.id || parentId,
  };
}

export function retainThreadContextMessages(
  threadContextHistory: TlonHistoryEntry[],
  maxMessages = MAX_THREAD_CONTEXT_MESSAGES
): TlonHistoryEntry[] {
  if (threadContextHistory.length <= maxMessages) {
    return threadContextHistory;
  }
  return [
    threadContextHistory[0],
    ...threadContextHistory.slice(-(maxMessages - 1)),
  ];
}

export function buildThreadContextMessage(
  threadContextHistory: TlonHistoryEntry[],
  currentMessageText: string,
  opts: {
    formatAuthor: (author: string) => string;
    sanitizeContent: (content: string) => string;
    maxMessages?: number;
  }
): { messageText: string; contextMessages: TlonHistoryEntry[] } | null {
  if (threadContextHistory.length === 0) {
    return null;
  }

  const contextMessages = retainThreadContextMessages(
    threadContextHistory,
    opts.maxMessages ?? MAX_THREAD_CONTEXT_MESSAGES
  );
  const threadContext = contextMessages
    .map(
      (msg) =>
        `${opts.formatAuthor(msg.author)}: ${opts.sanitizeContent(renderHistoryContent(msg))}`
    )
    .join('\n');
  const contextNote = `[Thread conversation - ${contextMessages.length} messages including the parent post. You are participating in this thread. Only respond if relevant or helpful - you don't need to reply to every message.]`;
  return {
    messageText:
      `${contextNote}\n\n[Previous messages]\n${threadContext}\n\n` +
      `[Current message]\n${currentMessageText}`,
    contextMessages,
  };
}

/**
 * Assemble complete thread context with parent post first, then replies.
 */
export async function fetchThreadContextHistory(
  api: { scry: (path: string) => Promise<unknown> },
  channelNest: string,
  parentId: string,
  count = 50,
  runtime?: RuntimeEnv
): Promise<TlonHistoryEntry[]> {
  const [parentPost, replies] = await Promise.all([
    fetchParentPostHistoryEntry(api, channelNest, parentId, runtime),
    fetchThreadHistory(api, channelNest, parentId, count, runtime),
  ]);

  const ordered = [parentPost, ...replies].filter(
    (entry): entry is TlonHistoryEntry => Boolean(entry)
  );
  const seen = new Set<string>();
  const deduped: TlonHistoryEntry[] = [];

  for (const entry of ordered) {
    const key = entry.id
      ? normalizeMessageId(entry.id)
      : `${entry.author}:${entry.timestamp}:${entry.content}`;
    if (seen.has(key)) {
      continue;
    }
    seen.add(key);
    deduped.push(entry);
  }

  return deduped;
}

File: packages/hermes-tlon-adapter/history.py
"""Channel and thread history context for group dispatches.

Mirrors the OpenClaw Tlon plugin's context enrichment: when the bot wakes in a
group channel it gets recent channel activity (or thread history for thread
replies) prepended to the message text, so it can answer in context instead of
seeing one isolated line.

Reads use the same `/channels/v4` scry surface as OpenClaw.
"""

from __future__ import annotations

import asyncio
import logging
import re
from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Optional, Sequence

from .media import render_content_with_blob
from .tlon_api import extract_message_text, normalize_ship

logger = logging.getLogger(__name__)

ScryFn = Callable[[str], Awaitable[Any]]

# Neutralize role tags so history content cannot impersonate envelope framing.
_ROLE_TAG_RE = re.compile(r"\[(owner|user|admin|system)\]", re.IGNORECASE)


def sanitize_context_text(text: str) -> str:
    return _ROLE_TAG_RE.sub(r"(\1)", str(text or ""))


def format_ud(post_id: Any) -> str:
    """Format a post id as @ud with dots every 3 digits from the right."""
    digits = str(post_id or "").replace(".", "")
    if not digits:
        return ""
    chunks: list[str] = []
    for end in range(len(digits), 0, -3):
        chunks.append(digits[max(0, end - 3) : end])
    return ".".join(reversed(chunks))


def _normalize_id(post_id: Any) -> str:
    return str(post_id or "").replace(".", "")


@dataclass(frozen=True)
class HistoryEntry:
    author: str
    content: str
    timestamp: float
    post_id: str = ""
    blob: Optional[str] = None


def _entry_from_content(content: Any, seal: Any, fallback_id: Any = None) -> Optional[HistoryEntry]:
    if not isinstance(content, dict):
        return None
    text = extract_message_text(content.get("content"))
    raw_blob = content.get("blob")
    blob = raw_blob if isinstance(raw_blob, str) and raw_blob.strip() else None
    if not text.strip() and not blob:
        return None
    post_id = ""
    if isinstance(seal, dict):
        post_id = str(seal.get("id") or "")
    if not post_id and fallback_id is not None:
        post_id = str(fallback_id or "")
    try:
        timestamp = float(content.get("sent") or 0)
    except (TypeError, ValueError):
        timestamp = 0.0
    return HistoryEntry(
        author=normalize_ship(str(content.get("author") or "")) or "unknown",
        content=text,
        timestamp=timestamp,
        post_id=post_id,
        blob=blob,
    )


def parse_channel_history(payload: Any) -> list[HistoryEntry]:
    if isinstance(payload, list):
        posts: Sequence[Any] = payload
    elif isinstance(payload, dict):
        raw_posts = payload.get("posts")
        posts = list(raw_posts.values()) if isinstance(raw_posts, dict) else list(payload.values())
    else:
        return []

    entries: list[HistoryEntry] = []
    for item in posts:
        if not isinstance(item, dict):
            continue
        post_set = item.get("r-post", {}).get("set") if isinstance(item.get("r-post"), dict) else None
        essay = item.get("essay") or (post_set or {}).get("essay")
        seal = item.get("seal") or (post_set or {}).get("seal")
        entry = _entry_from_content(essay, seal)
        if entry is not None:
            entries.append(entry)
    entries.sort(key=lambda entry: entry.timestamp)
    return entries


def parse_thread_replies(payload: Any) -> list[HistoryEntry]:
    if isinstance(payload, list):
        replies: Sequence[Any] = payload
    elif isinstance(payload, dict):
        raw_replies = payload.get("replies")
        if isinstance(raw_replies, list):
            replies = raw_replies
        elif isinstance(raw_replies, dict):
            replies = list(raw_replies.values())
        else:
            replies = list(payload.values())
    else:
        return []

    entries: list[HistoryEntry] = []
    for item in replies:
        if not isinstance(item, dict):
            continue
        reply_set = item.get("r-reply", {}).get("set") if isinstance(item.get("r-reply"), dict) else None
        memo = item.get("memo") or (reply_set or {}).get("memo")
        seal = item.get("seal") or (reply_set or {}).get("seal")
        entry = _entry_from_content(memo, seal, fallback_id=item.get("id"))
        if entry is not None:
            entries.append(entry)
    entries.sort(key=lambda entry: entry.timestamp)
    return entries


def parse_parent_post(payload: Any, parent_id: str) -> Optional[HistoryEntry]:
    if not isinstance(payload, dict):
        return None
    post = payload.get("post") if isinstance(payload.get("post"), dict) else payload
    post_set = post.get("r-post", {}).get("set") if isinstance(post.get("r-post"), dict) else None
    content = post.get("essay") or post.get("memo") or (post_set or {}).get("essay")
    seal = post.get("seal") or (post_set or {}).get("seal")
    return _entry_from_content(content, seal, fallback_id=parent_id)


async def fetch_channel_history(scry: ScryFn, nest: str, count: int) -> list[HistoryEntry]:
    payload = await scry(f"/channels/v4/{nest}/posts/newest/{count}/outline")
    return parse_channel_history(payload)


async def fetch_thread_context(
    scry: ScryFn,
    nest: str,
    parent_id: str,
    count: int,
) -> list[HistoryEntry]:
    """Parent post first, then replies oldest-first, deduplicated."""
    parent_ud = format_ud(parent_id)

    async def fetch_parent() -> Optional[HistoryEntry]:
        try:
            payload = await scry(f"/channels/v4/{nest}/posts/post/{parent_ud}")
        except Exception as exc:
            logger.debug("[tlon] parent post fetch failed for %s: %s", parent_id, exc)
            return None
        return parse_parent_post(payload, parent_id)

    async def fetch_replies() -> list[HistoryEntry]:
        try:
            payload = await scry(
                f"/channels/v4/{nest}/posts/post/id/{parent_ud}/replies/newest/{count}"
            )
        except Exception as exc:
            logger.debug("[tlon] thread replies fetch failed for %s: %s", parent_id, exc)
            return []
        return parse_thread_replies(payload)

    parent, replies = await asyncio.gather(fetch_parent(), fetch_replies())
    ordered = ([parent] if parent else []) + replies
    seen: set[str] = set()
    deduped: list[HistoryEntry] = []
    for entry in ordered:
        key = (
            _normalize_id(entry.post_id)
            or f"{entry.author}:{entry.timestamp}:{entry.content}:{entry.blob or ''}"
        )
        if key in seen:
            continue
        seen.add(key)
        deduped.append(entry)
    return deduped


def render_history_content(entry: HistoryEntry) -> str:
    return render_content_with_blob(entry.content, entry.blob, compact=True)


def _context_line_items(entries: Sequence[HistoryEntry]) -> list[tuple[HistoryEntry, str]]:
    items: list[tuple[HistoryEntry, str]] = []
    for entry in entries:
        rendered = sanitize_context_text(render_history_content(entry))
        if rendered.strip():
            items.append((entry, rendered))
    return items


def _context_lines(entries: Sequence[HistoryEntry]) -> str:
    return "\n".join(
        f"{entry.author}: {rendered}" for entry, rendered in _context_line_items(entries)
    )


def build_channel_context(
    entries: Sequence[HistoryEntry],
    *,
    current_text: str,
    current_id: str,
    is_mention: bool,
    limit: int,
) -> Optional[str]:
    """Prepend recent channel activity to the current message text."""
    if limit <= 0:
        return None
    current_key = _normalize_id(current_id)
    context_entries = [
        entry
        for entry in entries
        if not current_key or _normalize_id(entry.post_id) != current_key
    ]
    context = _context_line_items(context_entries)[-limit:]
    if not context:
        return None

    note = (
        f"[Recent channel activity - {len(context)} messages. "
        "Use this context to understand what's being discussed.]"
    )
    current_label = "[Current message (mentioned you)]" if is_mention else "[Current message]"
    lines = "\n".join(f"{entry.author}: {rendered}" for entry, rendered in context)
    return f"{note}\n\n{lines}\n\n{current_label}\n{current_text}"


def build_thread_context(
    entries: Sequence[HistoryEntry],
    *,
    current_text: str,
    current_id: str,
    limit: int,
) -> Optional[str]:
    """Prepend thread history (parent post first) to the current message text."""
    if limit <= 0:
        return None
    current_key = _normalize_id(current_id)
    context_entries = [
        entry
        for entry in entries
        if not current_key or _normalize_id(entry.post_id) != current_key
    ]
    context = _context_line_items(context_entries)
    if not context:
        return None
    if len(context) > limit:
        # limit == 1 keeps only the parent: [-(limit - 1):] would be [-0:],
        # which is the whole list.
        context = [context[0]] if limit == 1 else [context[0], *context[-(limit - 1) :]]

    note = (
        f"[Thread conversation - {len(context)} messages including the parent post. "
        "You are participating in this thread. Only respond if relevant or helpful - "
        "you don't need to reply to every message.]"
    )
    lines = "\n".join(f"{entry.author}: {rendered}" for entry, rendered in context)
    return (
        f"{note}\n\n[Previous messages]\n{lines}\n\n"
        f"[Current message]\n{current_text}"
    )
